internal/dispatch/singleflight/singleflight.go

```diff
@@ -0,0 +1,86 @@
+package singleflight
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"resenje.org/singleflight"
+
+	"github.com/authzed/spicedb/internal/services/shared"
+
+	"github.com/authzed/spicedb/internal/dispatch"
+	v1 "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+)
+
+var singleFlightCount = promauto.NewCounterVec(prometheus.CounterOpts{
+	Namespace: "spicedb",
+	Subsystem: "dispatch",
+	Name:      "single_flight_total",
+	Help:      "total number of dispatch requests that were single flighted",
+}, []string{"shared"})
+
+func New(delegate dispatch.Dispatcher) dispatch.Dispatcher {
+	return &Dispatcher{delegate: delegate}
+}
+
+type Dispatcher struct {
+	delegate   dispatch.Dispatcher
+	checkGroup singleflight.Group[string, *v1.DispatchCheckResponse]
+}
+
+func (d *Dispatcher) DispatchCheck(ctx context.Context, req *v1.DispatchCheckRequest) (*v1.DispatchCheckResponse, error) {
+	key, err := hashForDispatchCheck(req)
```

Comment:
only doing dispatchcheck as a test? or is there some reason to not singleflight the rest?

Reply:
nvm, you mentioned this in the PR description

Reply:
There is no reason not to singleflight everything, I just wanted to get this out of the door ASAP and then we can follow up with singleflight for the rest of requests.